### PR TITLE
Add nav footer for cfr changes.

### PR DESCRIPTION
**Pull request by jmcarp**
[Resolves https://github.com/eregs/notice-and-comment/issues/196]

### CODE REVIEW

File: regulations/tests/views_preamble_tests.py [hunk 1]
```diff
@@ -208,3 +212,35 @@ def test_add_amendment(self, fetch_meta, fetch_toc):
                 title='99', part='222', name='Some title for reg 222',
                 authority_url='/preamble/docdoc/cfr_changes/222',
                 sections=[])])
+
+    def test_navigation(self):
```
Comment (cmc333333) [inline regulations/tests/views_preamble_tests.py]: This method will override the previous

Comment (jmcarp) [inline regulations/tests/views_preamble_tests.py]: It shouldn't--they're in different classes.

File: regulations/tests/views_preamble_tests.py [hunk 2]
```diff
@@ -208,3 +212,35 @@ def test_add_amendment(self, fetch_meta, fetch_toc):
...
+                        title=u'§ 478.120 Firearms',
+                        section='120',
+                    ),
+                ],
+            )
+        ]
+
+        nav = preamble.cfr_section_navigation('478-99', toc)
+        assert_equal(nav['next'].section_id, '2016_02749-cfr-478-120')
+        assert_is_none(nav['previous'])
```
Comment (cmc333333) [inline regulations/tests/views_preamble_tests.py]: There _should_ be an authority entry. It's the first "section" under 478

File: regulations/views/preamble.py [hunk 3]
```diff
@@ -202,6 +220,25 @@ def from_section(cls, sect):
             sub_label=label,
         )
 
+    cfr_re = re.compile(r'(§ [\d.]+) (.*)')
```
Comment (cmc333333) [inline regulations/views/preamble.py]: Does this need to be unicode?

File: regulations/tests/views_preamble_tests.py (hunk 1 above)
Comment (cmc333333) [inline regulations/tests/views_preamble_tests.py]: Oop, sorry, didn't expand the diff

**Comment by cmc333333** [comment]
This is a good start. I noted a few minor issues, but there are two major ones:
1) Users should be able to "Next"/"Prev" in between CFR parts
2) Users should be able to "Next"/"Prev" _between_ the Preamble and CFR changes

Did I get that right, @jehlers ?

File: regulations/views/preamble.py (hunk 3 above)
Comment (jmcarp) [inline regulations/views/preamble.py]: Shouldn't--we're importing `unicode_literals` at the top of the time.

File: regulations/tests/views_preamble_tests.py (hunk 2 above)
Comment (jmcarp) [inline regulations/tests/views_preamble_tests.py]: Sorry, not sure what this means.

File: regulations/views/preamble.py (hunk 3 above)
Comment (cmc333333) [inline regulations/views/preamble.py]: Very nice!

File: regulations/tests/views_preamble_tests.py (hunk 2 above)
Comment (cmc333333) [inline regulations/tests/views_preamble_tests.py]: 478.99 wouldn't ever be the "first" section; we'll always have an "Authority":
screen shot 2016-04-26 at 2 55 46 pm

**Comment by jmcarp** [comment]
@cmc333333: just so I understand, we're saying we want the first `ToCSection` in a `ToCPart` to be considered adjacent to its parent? And the `ToCPart` should be considered adjacent to the last `ToCSection` of the previous `ToCPart`?

**Comment by cmc333333** [comment]
The first part doesn't sound quite right, but the second sounds good.

Say we have a preamble ToC like so:
```
General Stuff
I. Something
    A. Something
    B. Something
II. Something
```
And CFR changes like:
```
11 CFR 222
    Authority
    222.33
    222.44
11 CFR 333
    Authority
    333.55
```

For the purpose of the footer links, the following should be considered a single list:
```
General Stuff
I. Something
II. Something
Authority (of 11 CFR 222)
222.33
222.44
Authority (of 11 CFR 333)
333.55
```

Does that make sense?

**Comment by jmcarp** [comment]
To clarify, the authority lives in a `ToCPart` tuple, and the sections live in its `sections` list of `ToCSect` tuples. We flatten that structure out in the toc markup so that the `ToCPart` appears to be adjacent to its first child. That's what I'm claiming that we want the first `ToCSect` to be treated as if it's adjacent to its containing `ToCPart`. Thoughts?

**Comment by cmc333333** [comment]
I think we're talking about two views of the same data. You're describing a parent-child relationship of peers, but the code's not written that way; notably, `authority` is intentionally separate from the `sections` as they contain different types of content. If making the `authority` field look like the entries in `sections` will make your task easier, go for it. As long as the end result is the same (functionality as described above), I don't think there's a huge difference.

**Comment by jmcarp** [comment]
Updated.

File: regulations/static/regulations/js/source/views/main/section-footer-view.js [hunk 4]
```diff
@@ -20,12 +20,14 @@ var SectionFooterView = Backbone.View.extend({
     },
 
     sendNavEvent: function(e) {
-        e.preventDefault();
         var $target = $(e.currentTarget);
         var sectionId = $target.data('linked-section');
         var pageType = $target.data('page-type') || 'reg-section';
 
-        MainEvents.trigger('section:open', sectionId, {}, pageType);
+        if (sectionId && pageType) {
```
Comment (cmc333333) [inline regulations/static/regulations/js/source/views/main/section-footer-view.js]: Minor, but isn't `pageType` guaranteed due to (left) line 26?

File: regulations/views/preamble.py [hunk 5]
```diff
@@ -172,21 +178,69 @@ def make_preamble_toc(nodes, depth=1, max_depth=3):
...
+    pairs = list(CfrChangePair.from_toc(cfr_toc))
+    for idx, pair in enumerate(pairs):
+        if pair.match(part, section):
+            if idx > 0:
+                nav['previous'] = NavItem.from_object(pairs[idx - 1].item)
+            else:
+                nav['previous'] = NavItem.from_object(preamble_toc[-1])
+            if idx < len(pairs) - 1:
+                nav['next'] = NavItem.from_object(pairs[idx + 1].item)
+            break
```
Comment (cmc333333) [inline regulations/views/preamble.py]: This `break` is superfluous, yes?

File: regulations/views/preamble.py [hunk 6]
```diff
@@ -172,21 +178,69 @@ def make_preamble_toc(nodes, depth=1, max_depth=3):
...
+                    preamble_toc[idx - 1])
+            if idx < len(preamble_toc) - 1:
+                nav['next'] = NavItem.from_preamble_section(
+                    preamble_toc[idx + 1])
+            elif cfr_toc:
+                nav['next'] = NavItem.from_object(cfr_toc[0])
+    return nav
+
+
+def cfr_section_navigation(label_id, preamble_toc, cfr_toc):
```
Comment (cmc333333) [inline regulations/views/preamble.py]: Doesn't block this PR, but there's a lot of overlap between `cfr_section_navigation` and `preamble_section_navigation`, hey? Why not generate your list of `NavItems` once and step through the combined list in a single function?

File: regulations/views/preamble.py [hunk 7]
```diff
@@ -172,21 +178,69 @@ def make_preamble_toc(nodes, depth=1, max_depth=3):
...
+        return self.match_part(part) and self.match_sect(sect)
+
+    def match_part(self, part):
+        return self.part == part
+
+    def match_sect(self, sect):
+        return (
+            sect is None or
+            (isinstance(self.item, ToCSect) and self.item.section == sect)
+        )
```
Comment (cmc333333) [inline regulations/views/preamble.py]: Please document this logic. It's not clear by `None` should "match" a `ToCSect`

**Comment by cmc333333** [comment]
The docs are the only piece that's really necessary, but I recommend taking a step back and trying to approach the problem again knowing what you've now run into. I get the sense from this that you're trying to maintain existing functions/objects rather than refactor, but it's hurting the readability.

File: regulations/views/preamble.py (hunk 7 above)
Comment (jmcarp) [inline regulations/views/preamble.py]: Will do. Note that `self.item` isn't necessarily a `ToCSect`.

**Comment by jmcarp** [comment]
Updated to consolidate some code--hopefully this is clearer now. I'm expecting much of the logic behind preamble and cfr change tocs to get moved to -parser and made more uniform, which should improve things more. @cmc333333: if you have a different approach in mind, I think it will be quicker to discuss on chat.

File: regulations/views/preamble.py [hunk 8]
```diff
@@ -52,9 +58,50 @@ def generate_html_tree(subtree, request, id_prefix=None):
...
+)
+
+
+class ToCPart(namedtuple('ToCPart',
+              ['title', 'part', 'name', 'authority_url', 'sections'])):
+    def to_nav_item(self):
+        return NavItem(
+            url=self.authority_url,
+            section_id='',
+            markup_prefix='Authority',
```
Comment (cmc333333) [inline regulations/views/preamble.py]: Small detail but would 
```python
markup_prefix="{} CFR {}".format(self.title, self.part)
```
be a good chunk of text here?

**Comment by cmc333333** [comment]
Works like a charm locally and the code is much easier to follow now. Added one small request, but if you don't have time to get to it, I'm ready to merge.

**Pull request merged**